File: rocket_c/src/summaries.c
#include "summaries.h"

#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#include "logger.h"
#include "lora_transport.h"
#include "sensors.h"
#include "util.h"

static void append_name(char *buffer, size_t len, const char *name, bool *first) {
    if (*first) {
        snprintf(buffer, len, "%s", name);
        *first = false;
    } else {
        strncat(buffer, ", ", len - strlen(buffer) - 1);
        strncat(buffer, name, len - strlen(buffer) - 1);
    }
}

void log_start_summary(SensorCaps caps) {
    char activos[128] = "";
    char inactivos[128] = "";
    bool first_act = true;
    bool first_inact = true;

    if (caps.has_mpu) {
        append_name(activos, sizeof(activos), "mpu6050", &first_act);
    } else {
        append_name(inactivos, sizeof(inactivos), "mpu6050", &first_inact);
    }
    if (caps.has_bmp) {
        append_name(activos, sizeof(activos), "bmp180", &first_act);
    } else {
        append_name(inactivos, sizeof(inactivos), "bmp180", &first_inact);
    }
    if (caps.has_gps) {
        append_name(activos, sizeof(activos), "neo6m", &first_act);
    } else {
        append_name(inactivos, sizeof(inactivos), "neo6m", &first_inact);
    }

    const char *activos_txt = first_act ? "ninguno" : activos;
    const char *inactivos_txt = first_inact ? "ninguno" : inactivos;

    log_message(LOG_LEVEL_SYS, "SYSTEM", "===== RESUMEN INICIAL =====");
    log_message(LOG_LEVEL_INFO, "SYSTEM", "Sensores disponibles: %s", activos_txt);
    log_message(LOG_LEVEL_WARN, "SYSTEM", "Sensores NO disponibles: %s", inactivos_txt);
    log_message(LOG_LEVEL_INFO, "SYSTEM", "LoRa: %s", lora_is_ready() ? "LISTO" : "NO LISTO");
}

void log_final_summary(const ActivityTracker *tracker) {
    char reales[128] = "";
    char dummy[128] = "";
    char sin_datos[128] = "";
    bool first_reales = true;
    bool first_dummy = true;
    bool first_sin_datos = true;

    for (size_t i = 0; i < SENSOR_COUNT; ++i) {
        SensorType sensor = (SensorType)i;
        const char *name = sensor_name(sensor);
        if (!activity_tracker_seen(tracker, sensor)) {
            append_name(sin_datos, sizeof(sin_datos), name, &first_sin_datos);
        } else if (activity_tracker_last_dummy(tracker, sensor)) {
            append_name(dummy, sizeof(dummy), name, &first_dummy);
        } else {
            append_name(reales, sizeof(reales), name, &first_reales);
        }
    }

    const char *reales_txt = first_reales ? "ninguno" : reales;
    const char *dummy_txt = first_dummy ? "ninguno" : dummy;
    const char *sin_datos_txt = first_sin_datos ? "ninguno" : sin_datos;

    log_message(LOG_LEVEL_SYS, "SYSTEM", "===== RESUMEN FINAL =====");
    log_message(LOG_LEVEL_INFO, "SYSTEM", "Datos REALES recibidos: %s", reales_txt);
    log_message(LOG_LEVEL_WARN, "SYSTEM", "Datos DUMMY (sin hardware): %s", dummy_txt);
    log_message(LOG_LEVEL_ERROR, "SYSTEM", "Sensores sin datos: %s", sin_datos_txt);

    ZeroSignalInfo info = activity_tracker_zero_details(tracker);
    if (info.sent) {
        char ts[64];
        isoformat_utc(info.timestamp, ts, sizeof(ts));
        log_message(LOG_LEVEL_INFO, "SYSTEM", "Senal por aceleracion cero: ENVIADA (t=%s, |a|=%.3fg)",
                    ts, info.magnitude);
    } else {
        log_message(LOG_LEVEL_WARN, "SYSTEM", "Senal por aceleracion cero: NO ENVIADA");
    }
    log_message(LOG_LEVEL_INFO, "SYSTEM", "LoRa: %s", lora_is_ready() ? "LISTO" : "NO LISTO");
}

File: summaries.py
from __future__ import annotations

from pathlib import Path
import time
from typing import List, Tuple

from logger import log
from lora_transport import get_init_error, has_link_failure, is_ready
from sensor_workers import CAPS
from aggregator import ActivityTracker
from sensor_messages import isoformat_utc

_FINAL_LOG_FILE = Path(__file__).resolve().parent / "logs" / "resumen_final.log"

def log_start_summary() -> None:
    activos = [sensor for sensor, ok in CAPS.items() if ok]
    inactivos = [sensor for sensor, ok in CAPS.items() if not ok]
    log("SYSTEM", "===== RESUMEN INICIAL =====", "SYS")
    log("SYSTEM", f"Sensores disponibles: {', '.join(activos) if activos else 'ninguno'}", "INFO")
    log("SYSTEM", f"Sensores NO disponibles: {', '.join(inactivos) if inactivos else 'ninguno'}", "WARN")
    log("SYSTEM", f"LoRa: {'LISTO' if is_ready() else 'NO LISTO'}", "INFO")

def log_final_summary(tracker: ActivityTracker) -> None:
    records: List[Tuple[str, str]] = []

    def record(message: str, level: str) -> None:
        log("SYSTEM", message, level)
        records.append((level, message))

    header = "===== RESUMEN FINAL ====="
    log("SYSTEM", header, "SYS")
    grupos = tracker.groups()
    record(
        f"Datos REALES recibidos: {', '.join(grupos['reales']) if grupos['reales'] else 'ninguno'}",
        "INFO",
    )
    record(
        f"Datos DUMMY (sin hardware): {', '.join(grupos['dummy']) if grupos['dummy'] else 'ninguno'}",
        "WARN",
    )
    record(
        f"Sensores sin datos: {', '.join(grupos['sin_datos']) if grupos['sin_datos'] else 'ninguno'}",
        "ERROR",
    )
    zero_signal = tracker.zero_accel_signal_details()
    if zero_signal["sent"]:
        ts = zero_signal["timestamp"]
        ts_str = isoformat_utc(ts) if isinstance(ts, (int, float)) else "desconocido"
        magnitude = zero_signal["magnitude"]
        magnitude_str = f"{magnitude:.3f}g" if isinstance(magnitude, (int, float)) else "desconocido"
        record(f"Señal por aceleración cero: ENVIADA (t={ts_str}, |a|={magnitude_str})", "INFO")
    else:
        record("Señal por aceleración cero: NO ENVIADA", "WARN")
    lora_ready = is_ready()
    record(f"LoRa: {'LISTO' if lora_ready else 'NO LISTO'}", "INFO")
    if not lora_ready:
        error = get_init_error()
        detalle = error if error else "motivo desconocido"
        record(f"LoRa no se inició correctamente: {detalle}", "ERROR")
    if has_link_failure():
        record("LoRa no se conectó correctamente (módulo sin respuesta)", "ERROR")
    _persist_final_log(header, records)


def _persist_final_log(header: str, records: List[Tuple[str, str]]) -> None:
    timestamp = isoformat_utc(time.time())
    lines = [f"{header} ({timestamp})"]
    lines.extend(f"[{level}] {message}" for level, message in records)
    try:
        _FINAL_LOG_FILE.parent.mkdir(parents=True, exist_ok=True)
        with _FINAL_LOG_FILE.open("a", encoding="utf-8") as fp:
            fp.write("\n".join(lines) + "\n\n")
        log("SYSTEM", f"Resumen final guardado en {_FINAL_LOG_FILE}", "SYS")
    except OSError as exc:
        log("SYSTEM", f"No pude guardar resumen final en {_FINAL_LOG_FILE}: {exc}", "ERROR")
